windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+  PostInstallState = class(TState)
+  private
+      procedure HandleMSIInstallComplete;
+  public
+    procedure Enter; override;
+    procedure Exit; override;
+    procedure HandleCheck; override;
+    function  HandleKmShell : Integer; override;
+    procedure HandleDownload; override;
+    procedure HandleAbort; override;
+    procedure HandleInstallNow; override;
+    function StateName: string; override;
+  end;
+
+
+  { This class also controls the state flow see  }
+  TUpdateStateMachine = class
+  private
+    FForce: Boolean;
+    FAutomaticUpdate: Boolean;
+    FParams: TUpdateStateMachineParams;
+    FErrorMessage: string;
+    DownloadTempPath: string;
+    FShowErrors: Boolean;
+    FDownload: TUpdateStateMachineDownloadParams;
+
+    CurrentState: TState;
+    // State object for performance (could lazy create?)
+    FIdle: IdleState;
+    FUpdateAvailable: UpdateAvailableState;
+    FDownloading: DownloadingState;
+    FWaitingRestart: WaitingRestartState;
+    FInstalling: InstallingState;
+    FRetry: RetryState;
+    FPostInstall: PostInstallState;
+    function GetState: TStateClass;
+    procedure SetState(const Value: TStateClass);
+    procedure SetStateOnly(const Value: TStateClass);
+    function ConvertEnumState(const TEnumState: TUpdateState): TStateClass;
+
+    procedure ShutDown;
+    {
+      SavePackageUpgradesToDownloadTempPath saves any new package IDs to a
+      single file in the download tempPath. This procedure saves the IDs of any
+      new packages to a file named "upgrade_packages.inf" in the download
+      tempPath.
+    }
+    procedure SavePackageUpgradesToDownloadTempPath;
+    function checkUpdateSchedule : Boolean;
+
+    function SetRegistryState (Update : TUpdateState): Boolean;
+    function GetAutomaticUpdates: Boolean;
+    function SetApplyNow(Value : Boolean): Boolean;
+    function GetApplyNow: Boolean;
+
+  protected
+  property State: TStateClass read GetState write SetState;
+
+  public
+    constructor Create(AForce: Boolean);
+    destructor Destroy; override;
+
+    procedure   HandleCheck;
+    function    HandleKmShell : Integer;
+    procedure   HandleDownload;
+    procedure   HandleAbort;
+    procedure   HandleInstallNow;
+    function    CurrentStateName: string;
+
+    property ShowErrors: Boolean read FShowErrors write FShowErrors;
+    function CheckRegistryState : TUpdateState;
+
+  end;
+
+  IOnlineUpdateSharedData = interface
+    ['{7442A323-C1E3-404B-BEEA-5B24A52BBB0E}']
+    function Params: TUpdateStateMachineParams;
+  end;
+
+  TOnlineUpdateSharedData = class(TInterfacedObject, IOnlineUpdateSharedData)
+  private
+    FParams: TUpdateStateMachineParams;
+  public
+    constructor Create(AParams: TUpdateStateMachineParams);
+    function Params: TUpdateStateMachineParams;
+  end;
+  // Private Utility functions
+  function ConfigCheckContinue: Boolean;
+implementation
```

Comment:
```suggestion
end;

// Private Utility functions
function ConfigCheckContinue: Boolean;
  
implementation
```

If `ConfigCheckContinue` is a private utility function, shouldn't it be in the `implementation` section as a `forward` declaration?